Improve local compilation specific errors

pmvald opened the pull request:
As per request from customers, we need to have more clear and actionable error messages for local compilation.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald committed:
refactor(compiler-cli): upgrade error codes and messages for unresolved symbol errors in local compilation mode

A single error code is created to unify the common error pattern in local compilation mode where an imported const cannot be resolved, but needs to be resolved. This mainly happens for Angular decorator fields such as @Component.template.

The error messages are also upgraded to be more centered around this unifying theme.

pmvald committed:
refactor(compiler-cli): remove the trailing error message for LOCAL_COMPILATION_UNRESOLVED_CONST

The trailing error message comes from tracing the chain of DymaicValue which leads to a mostly useless error that highlights the same symbol as the original message and emits the error message "Unknown reference". This error message is removed in the favour of the original message which suffices.

pmvald committed:
fix(compiler-cli): highlight the unresolved element in the @Component.styles array for the error LOCAL_COMPILATION_UNRESOLVED_CONST

Currently the whole array is highlighted.

pmvald committed:
fix(compiler-cli): show the correct message for the error LOCAL_COMPILATION_UNRESOLVED_CONST when an unresolved symbol used for @Component.styles

Currently the correct error message is shown only if @Component.styles is an array with some unresolved element. This change supports the new case of string type for the @Component.styles field.

pmvald committed:
fix(compiler-cli): show specific error for unresolved @HostListener's event name in local compilation mode

Currently the error is a generic error "selector must be a string ...". This commit makes the error more specific to local compilation and adds some action items.

pmvald committed:
fix(compiler-cli): show specific error for unresolved @HostBinding's argument in local compilation mode

Currently the error is a generic error "selector must be a string ...". This commit makes the error more specific to local compilation and adds some action items.

pmvald committed:
fix(compiler-cli): show specific error for unresolved @Directive.exportAs in local compilation mode

    Currently the error is a generic error "exportAs must be a string ...". This commit makes the error more specific to local compilation and adds some action items.

devversion commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -405,6 +426,14 @@ export function extractHostBindings(
+
+            // Specific error for local compilation mode if the event name cannot be resolved
+            if (compilationMode === CompilationMode.LOCAL && resolved instanceof DynamicValue && resolved.isFromUnknownIdentifier()) {
+              throw new FatalDiagnosticError(
```
Almost feels like a helper could reduce lots of repetition of the error messages here.

jessicajaniuk reviewed (approved):
reviewed-for: public-api

pmvald committed:
refactor(compiler-cli): add a helper to generate error `LOCAL_COMPILATION_UNRESOLVED_CONST` in common scenarios

A helper `validateLocalCompilationUnresolvedConst` is added to encapsulate a common pattern which leads to the error `LOCAL_COMPILATION_UNRESOLVED_CONST`.

devversion commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -403,3 +404,18 @@ function getInheritedUndecoratedCtorDiagnostic(
 }
+
+/**
+ * Throws `FatalDiagnosticError` with error code `LOCAL_COMPILATION_UNRESOLVED_CONST` if the compilation mode is local and the value is not resolved due to being imported from external files. This is a common scenario for errors in local compilation mode, and so this helper can be used to quickly generate the relevant errors.
```
Nit: This comment should be wrapped.

devversion commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -403,3 +404,18 @@ function getInheritedUndecoratedCtorDiagnostic(
+ * 
+ * @param nodeToHighlight Node to be highlighted in teh error message. Will default to value.node if not provided.    
+ */
+export function validateLocalCompilationUnresolvedConst(compilationMode: CompilationMode, value: ResolvedValue, nodeToHighlight: ts.Node|null, errorMessage: string): void {
```
Nit: Consider using `assert` as it indicates more clearly that this will throw.

devversion reviewed (approved):
LGTM. Feel free to add merge ready when addressed

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -491,24 +506,21 @@ export function parseDirectiveStyles(
+      throw new FatalDiagnosticError(
+          ErrorCode.LOCAL_COMPILATION_UNRESOLVED_CONST, 
+          unresolvedNode, 
+          'Unresolved identifier found for @Component.styles field! Did you import this identifier from a file outside of the compilation unit? This is not allowed when Angular compiler runs in local mode. Possible solutions: 1) Move the declarations into a file within the compilation unit, 2) Inline the styles, 3) Move the styles into separate files and include it using @Component.styleUrls');
```
Done!

alxhub reviewed (approved):
Reviewed-for: public-api

jessicajaniuk commented:
This PR was merged into the repository by commit f39cb064183d984254bdf4e41b61d3dc9379738a.
